### PR TITLE
[improve] Allow user-specified S3 path

### PR DESCRIPTION
**Pull request by mukesh154**
### Motivation

To allow the user-specified path with the help of topics to path mappings from the user.
### Modifications

Added a mapping field to allow the user to specify topics to path mapping. Based on the mapping path, the topics data will be uploaded to s3.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Documentation

Check the box below.

Need to update docs?

- [ ] `doc-required`

  (If you need help on updating docs, create a doc issue)

- [x] `no-need-doc`

  (Please explain why)

- [ ] `doc`

  (If this PR contains doc changes)

### CODE REVIEW

File: src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java [hunk 1]
```diff
@@ -115,6 +122,21 @@ public void open(Map<String, Object> config, SinkContext sinkContext) throws Exc
         this.blobWriter = initBlobWriter(sinkConfig);
     }
 
+    private Map<String, String> parseTopicsToPathMappingString(String topicsToPathMappingString) {
+        Map<String, String> topicsToPathMapping = new HashMap<>();
+        if (StringUtils.isNotEmpty(topicsToPathMappingString)) {
+            for (String topicPathPair : topicsToPathMappingString.split(",")) {
+                String[] topicToPathMapping = topicPathPair.split("=");
+                if (topicToPathMapping.length == 2) {
+                    topicsToPathMapping.put(topicToPathMapping[0], topicToPathMapping[1]);
```
Comment (nicoloboschi) [inline src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java]: maybe it's better to print the final mapping here

File: src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java [hunk 2]
```diff
@@ -331,11 +353,21 @@ public String buildPartitionPath(Record<GenericRecord> message,
 
         String encodePartition = partitioner.encodePartition(message, partitioningTimestamp);
         String partitionedPath = partitioner.generatePartitionedPath(message.getTopicName().get(), encodePartition);
+        String trimmedTopicPath = partitionedPath.replace(PATH_SEPARATOR + encodePartition, StringUtils.EMPTY);
```
Comment (nicoloboschi) [inline src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java]: this is not good, undo things by default. This code runs many times. 
I'd suggest to move the logic inside the partitioner class, then

Also maybe we should keep a local HashMap<String,String> with topic name -> path (without the date) so we don't recompute again and again the same pattern

Comment (nicoloboschi) [reply src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java]: Also we need unit tests to check input / output with all the flags about the partitions and all the combinations

**Review by nicoloboschi — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java (hunk 1 above)
Comment (mukesh154) [reply src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java]: done

File: src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java (hunk 2 above)
Comment (mukesh154) [reply src/main/java/org/apache/pulsar/io/jcloud/sink/BlobStoreAbstractSink.java]: The logic moved to AbstractPartitioner with above requirements & Added unit tests as required

File: src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java [hunk 3]
```diff
@@ -42,35 +46,67 @@ public abstract class AbstractPartitioner<T> implements Partitioner<T> {
...
+                }
+            }
+        }
+        LOGGER.info("The final topics to path mapping is: {}", topicsToPathMapping);
+        return topicsToPathMapping;
     }
 
     @Override
     public String generatePartitionedPath(String topic, String encodedPartition) {
+        if (MapUtils.isNotEmpty(computedTopicsToPathMapping) && computedTopicsToPathMapping.containsKey(topic)) {
```
Comment (nicoloboschi) [inline src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java]: no need to check if empty, please remove

File: src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java [hunk 4]
```diff
@@ -42,35 +46,67 @@ public abstract class AbstractPartitioner<T> implements Partitioner<T> {
...
+            return StringUtils.join(Arrays.asList(computedTopicsToPathMapping.get(topic), encodedPartition),
```
Comment (nicoloboschi) [inline src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java]: this is super costly for nothing,
can't be just 
String.format("%s%s%s", mapped, separator, encodedPartition)

?

File: src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java [hunk 5]
```diff
@@ -42,35 +46,67 @@ public abstract class AbstractPartitioner<T> implements Partitioner<T> {
...
                 joinList.add(newTopicName.getLocalName());
-                joinList.add(Integer.toString(topicName.getPartitionIndex()));
-            } else {
-                joinList.add(topicName.getLocalName());
             }
-        } else {
-            TopicName newTopicName = TopicName.get(topicName.getPartitionedTopicName());
-            joinList.add(newTopicName.getLocalName());
+            String generatedTopicPath = StringUtils.join(joinList, PATH_SEPARATOR);
+            if (MapUtils.isNotEmpty(topicsToPathMapping) && topicsToPathMapping.containsKey(generatedTopicPath)) {
```
Comment (nicoloboschi) [inline src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java]: no need to check if empty, please remove

**Review by nicoloboschi — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java (hunk 4 above)
Comment (mukesh154) [reply src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java]: done

File: src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java (hunk 5 above)
Comment (mukesh154) [reply src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java]: removed

File: src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java (hunk 3 above)
Comment (mukesh154) [reply src/main/java/org/apache/pulsar/io/jcloud/partitioner/AbstractPartitioner.java]: removed

**Review by nicoloboschi — APPROVED** [review]

**Pull request merged**